SNOW-966003: Fix Arrow return value for zero-row queries

thomasaarholt opened the pull request:
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #1800

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   This PR copies the approach done in #1275 by @jakobgm, but based on the latest main (I'm sorry @jakobgm, but it was easier to just copy rather than rebase your solution, and I hope you don't mind that), with one more type hint added due to some later commits. That approach itself is based on the current implementation of `.fetch_pandas_all()`.

   This is a minor code change, copying methodology already implemented by snowflake on the pandas side, but it makes it a lot easier to work with arrow data from the snowflake-connector-python.

   Please let me know if anything more is required from me.

sfc-gh-aling commented:
can we add some test here?

thomasaarholt commented:
Thanks for the quick response. I've added a quick test to check that the dtype of a integer column in fact is what we expect (the highest bit-depth).

thomasaarholt commented:
I'm a little worried about seeing this PR go stale like #1275. Is there anything further that I can do to finish this?

sfc-gh-aling commented on src/snowflake/connector/cursor.py:
```diff
@@ -1117,7 +1117,7 @@ def fetch_arrow_batches(self) -> Iterator[Table]:
         return self._result_set._fetch_arrow_batches()
 
-    def fetch_arrow_all(self) -> Table | None:
+    def fetch_arrow_all(self) -> Table:
```
the change looks quite neat to me.

@sfc-gh-mkeller , do you think this is a API breaking change that previously we return `None` now we return an empty `Table`.

I'm thinking of introducing a parameter to protect the change

thomasaarholt replied:
**Some thoughts**
1. Here's a suggested fix using your ideas:
```python
from typing import overload, Literal
from warnings import warn
from pyarrow import Table


@overload
def fetch_arrow_all(return_table_even_no_data: Literal[False] = False) -> None:
    ...


@overload
def fetch_arrow_all(return_table_even_no_data: Literal[True] = True) -> Table:
    ...


def fetch_arrow_all(return_table_even_no_data: bool = False) -> Table | None:
    if not return_table_even_no_data:
        warn(
            "The default behaviour of fetch_arrow_all when the query returns zero rows "
            "will change from returning None to returning an empty pyarrow table with "
            "schema using the highest bit length in version X.X.X.",
            DeprecationWarning,
            stacklevel=2,
        )
    # regular code here


foo = fetch_arrow_all(return_table_even_no_data=True) # Table, or "Unknown" if no type stubs for pyarrow are present
bar = fetch_arrow_all(return_table_even_no_data=False) # None

```

I'm happy to change any of this.

2. You _could_ consider this a bug-fix though. It is a discrepancy between the pandas and pyarrow apis. Depends on how strict you want to be about potentially breaking changes.

3. If you follow semver, then I'd expect that you would only introduce breaking changes with major version change. As a user of the connector, it would be nice to know when one expects breaking changes, even if you don't follow semver.

sfc-gh-aling replied:
hi @thomasaarholt, thanks for the proposal, I personally like it. re your third point, yeah.. sometime I feel torn when facing a breaking change bug fix, but the fact is that there might be programs might depend on this "bug" behavior, so even with a subtle change something might go wrong.

so I feel it would be safer to introduce a parameter for the new behavior.

I'm thinking of a better naming `force_return_table` instead of the `return_table_even_no_data`.

sfc-gh-aling replied:
can you help make the code change based upon our discussion? really appreciate your contribution and the discussion!

thomasaarholt replied:
Done! I decided to omit the line containing the X.X.X version entirely.

thomasaarholt commented:
Changes are implemented. This is ready for review again.

sfc-gh-aling commented on src/snowflake/connector/result_set.py:
```diff
@@ -164,13 +164,13 @@ def _fetch_arrow_batches(
         return self._create_iter(iter_unit=IterUnit.TABLE_UNIT, structure="arrow")
 
-    def _fetch_arrow_all(self) -> Table | None:
+    def _fetch_arrow_all(self) -> Table:
```
you also need to make the code change here
```suggestion
    def _fetch_arrow_all(self, force_return_table: bool = False) -> Table:
```

sfc-gh-aling commented on src/snowflake/connector/result_set.py:
```diff
@@ -164,13 +164,13 @@ def _fetch_arrow_batches(
             return pa.concat_tables(tables)
         else:
-            return None
+            return self.batches[0].to_arrow()
```
```suggestion
            return self.batches[0].to_arrow() if force_return_table else None
```

sfc-gh-aling commented on src/snowflake/connector/cursor.py:
```diff
@@ -1117,8 +1118,25 @@ def fetch_arrow_batches(self) -> Iterator[Table]:
+        ...
+
+    def fetch_arrow_all(self, force_return_table: bool = False) -> Table | None:
+        if not force_return_table:
+            warn(
+                "The behaviour of fetch_arrow_all when the query returns zero rows "
+                "will change from returning None to returning an empty pyarrow table with "
+                "schema using the highest bit length for each column.",
+                DeprecationWarning,
+                stacklevel=2,
+            )
```
let's add a docstring to the method instead of a warning in run time
```suggestion
        if not force_return_table:
            warn(
                "The behaviour of fetch_arrow_all when the query returns zero rows "
                "will change from returning None to returning an empty pyarrow table with "
                "schema using the highest bit length for each column.",
                DeprecationWarning,
                stacklevel=2,
            )
```

sfc-gh-aling commented on src/snowflake/connector/cursor.py:
```diff
@@ -1117,8 +1118,25 @@ def fetch_arrow_batches(self) -> Iterator[Table]:
+                stacklevel=2,
+            )
         self.check_can_use_arrow_resultset()
+
```
line 1145 shall be updated to be 
`return self._result_set._fetch_arrow_all(force_return_table= force_return_table)`

sfc-gh-aling commented on test/integ/pandas/test_arrow_pandas.py:
```diff
@@ -1167,6 +1167,17 @@ def test_simple_arrow_fetch(conn_cnx):
+def test_arrow_zero_rows(conn_cnx):
+    with conn_cnx() as cnx:
+        with cnx.cursor() as cur:
+            cur.execute(SQL_ENABLE_ARROW)
+            cur.execute("select 1::NUMBER(38,0) limit 0")
+            table = cur.fetch_arrow_all()
+            # Snowflake will return an integery dtype with maximum bit-depth if
+            # no rows are returned
+            assert table.schema[0].type == pyarrow.int64()
```
```suggestion
            cur.execute(SQL_ENABLE_ARROW)
            cur.execute("select 1::NUMBER(38,0) limit 0")
            table = cur.fetch_arrow_all(force_return_table=True)
            # Snowflake will return an integery dtype with maximum bit-depth if
            # no rows are returned
            assert table.schema[0].type == pyarrow.int64()
            cur.execute("select 1::NUMBER(38,0) limit 0")
            # test default behavior
            assert cur.fetch_arrow_all(force_return_table=False) is None
```

sfc-gh-aling commented on src/snowflake/connector/cursor.py:
```diff
@@ -1117,8 +1118,25 @@ def fetch_arrow_batches(self) -> Iterator[Table]:
+    def fetch_arrow_all(self, force_return_table: Literal[True] = True) -> Table:
+        ...
+
+    def fetch_arrow_all(self, force_return_table: bool = False) -> Table | None:
```
we shall also update the changelog

sfc-gh-aling commented:
thanks! @thomasaarholt I did another round of review and left some comments.

thomasaarholt commented:
Thank you! I can't believe I didn't actually implement it...! My two-year-old is making me sleep-deprived. I'm very grateful for your code suggestions.

I've incorporated them and updated the docstring with a bit more info and also overloaded the `_fetch_arrow_all` method. I'm unable to make the tests run locally, but I think it looks good.

thomasaarholt commented:
Btw, I would encourage you to change the [pyarrow import here](https://github.com/snowflakedb/snowflake-connector-python/blob/aa5e50b9004ad38a3607a0e70dfd3445a501457c/src/snowflake/connector/cursor.py#L84-L88) so that you don't set `Table = None`, and instead use `if TYPE_CHECKING`, since `Table` is only used as a type, not directly called anywhere. Pyarrow typing is pretty non-existant, but this would remove a lot of red error messages in the function definitions.

I considered doing so in this PR, but decided to avoid scope creep 😅

sfc-gh-aling commented:
hey @thomasaarholt , I like your proposal on the `Table` , are you willing to make another PR for it?

for this PR, I tried to do some small changes on your PR, but looks like I don't have the permission to push to your branch..
```
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
To github.com:thomasaarholt/snowflake-connector-python.git
 ! [remote rejected]   thomasaarholt/thomasaarholt/fix-arrow-empty-rows -> thomasaarholt/thomasaarholt/fix-arrow-empty-rows (permission denied)
error: failed to push some refs to 'github.com:thomasaarholt/snowflake-connector-python.git'
```

I added another two commits which tried to fix lint, update doc: https://github.com/snowflakedb/snowflake-connector-python/pull/1832.

would you be good if I merge the other PR to avoid you going back and forth tweaking the minor changes?

thomasaarholt commented:
That’s fine! Go right ahead! I’ll be happy to make a PR on the Table tomorrow (evening in Europe now). 

I’ll close this when you merge the other one!

sfc-gh-aling commented:
sounds good @thomasaarholt , I will make sure the PR is merged this week.
appreciate your contribution!

thomasaarholt commented:
This PR was merged via #1832! Closing.
